File: backend-py/app/form_pipeline/codegen/build_ff_js.py
"""Port of packages/shared/src/codegen/js/buildFfJs.ts.

`<Sub>-<lang>_<projectCode>_FF.js` is the reference's `SGE-EN_F2H26_FF.js`,
byte-identical — no per-form logic — preceded only by the one-line
`language_pin` that makes this language the page's default. One file per
language (the form's default language when `locale` is omitted).
"""

from __future__ import annotations

from typing import Optional

from .file_names import FileNames, language_file_names
from .language_pin import language_pin
from .reference_ff_js import REFERENCE_FF_JS
from .types import GeneratedFile


def build_ff_js(file_names: FileNames, locale: Optional[str] = None) -> GeneratedFile:
    files = language_file_names(file_names, locale)
    return GeneratedFile(path=files.ffJs, contents=language_pin(files.locale) + REFERENCE_FF_JS)

File: packages/shared/src/codegen/js/buildFfJs.ts
import type { LocaleCode } from "../../form/formDefinition";
import { languageFileNames, type FileNames } from "../fileNames";
import type { GeneratedFile } from "../types";
import { languagePin } from "./languagePin";
import { REFERENCE_FF_JS } from "./referenceFfJsContent";

/**
 * `<Sub>-<lang>_<projectCode>_FF.js` is the reference's `SGE-EN_F2H26_FF.js`, byte-identical
 * — no per-form logic — preceded only by the one-line `languagePin` that makes this
 * language the page's default. It reads the bare globals `fields`/`questions`/`answers`/
 * `page_error`/`validation_messages`/`param`/`country_subsidiary`/`subsidiary_detail`
 * that `buildDataJs.ts` emits, and the DOM shape `buildFfHtml.ts` renders. One file per
 * language (the form's default language when `locale` is omitted).
 */
export function buildFfJs(fileNames: FileNames, locale?: LocaleCode): GeneratedFile {
  const files = languageFileNames(fileNames, locale);
  return { path: files.ffJs, contents: languagePin(files.locale) + REFERENCE_FF_JS };
}
